storage/internal/elastic/utils.go

```diff
@@ -0,0 +1,161 @@
+var storeTypeAndDeploymentIDRegex = regexp.MustCompile(`(?m)\_yorc\/(\w+)\/?(.+)?\/?`)
+
+// Parse a key of form "_yorc/events/" or "_yorc/logs/MyApp" or "_yorc/logs/MyApp/" to get the store type (logs|events) and eventually the deploymentId.
+func extractStoreTypeAndDeploymentID(k string) (storeType string, deploymentID string) {
+	res := storeTypeAndDeploymentIDRegex.FindAllStringSubmatch(k, -1)
+	for i := range res {
+		storeType = res[i][1]
+		if len(res[i]) == 3 {
+			deploymentID = res[i][2]
+			if strings.HasSuffix(deploymentID, "/") {
+				deploymentID = deploymentID[:len(deploymentID)-1]
+			}
+		}
+	}
+	return storeType, deploymentID
+}
+
+// We need to append JSON directly into []byte to avoid useless and costly  marshaling / unmarshaling.
+func appendJSONInBytes(a []byte, v []byte) []byte {
+	last := len(a) - 1
+	lastByte := a[last]
+	// just append v at the end
+	a = append(a, v...)
+	// then slice
+	for i, s := range v {
+		a[last+i] = s
+	}
+	a[last+len(v)] = lastByte
+	return a
+}
+
+func parseInt64StringToUint64(value string) (uint64, error) {
+	valueInt, err := strconv.ParseInt(value, 10, 64)
+	if err != nil {
+		return 0, errors.Wrapf(err, "Not able to parse string: %s to int64, error was: %+v", value, err)
+	}
+	result := uint64(valueInt)
+	return result, nil
+}
+
+// The max uint is 9223372036854775807 (19 cars), this is the maximum nano for a time (2262-04-12 00:47:16.854775807 +0100 CET).
+// Since we use nanotimestamp as ID for events and logs, and since we store this ID as string in ES index, we must ensure that
+// the string will be comparable.
+func getSortableStringFromUint64(nanoTimestamp uint64) string {
+	nanoTimestampStr := strconv.FormatUint(nanoTimestamp, 10)
+	if len(nanoTimestampStr) < 19 {
+		nanoTimestampStr = strings.Repeat("0", 19-len(nanoTimestampStr)) + nanoTimestampStr
+	}
+	return nanoTimestampStr
+}
+
+// The document is enriched by adding 'clusterId' and 'iid' properties.
+// This addition is done by directly manipulating the []byte in order to avoid costly successive marshal / unmarshal operations.
+func buildElasticDocument(clusterID string, k string, v interface{}) (string, []byte, error) {
+	// Extract indice name and timestamp by parsing the key
+	storeType, timestamp := extractStoreTypeAndTimestamp(k)
+	log.Debugf("storeType is: %s, timestamp: %s", storeType, timestamp)
+
+	// Convert timestamp to an int64
+	eventDate, err := time.Parse(time.RFC3339Nano, timestamp)
+	if err != nil {
+		return storeType, nil, errors.Wrapf(err, "failed to parse timestamp %+v as time, error was: %+v", timestamp, err)
+	}
+	// Convert to UnixNano int64
+	iid := eventDate.UnixNano()
+
+	// This is the piece of 'JSON' we want to append
+	a := `,"iid":"` + strconv.FormatInt(iid, 10) + `","clusterId":"` + clusterID + `"`
+	// v is a json.RawMessage
+	raw := v.(json.RawMessage)
+	raw = appendJSONInBytes(raw, []byte(a))
```

Comment:
json.RawMessage is an array of bytes So you don't need to do this ?

Reply:
I do this to avoid marshal / unmarshall operations. I just need to append this piece of JSON in a a []byte. json.RawMessage is actually a []byte. How could I do this in another way ?

Reply:
My mistake, I misunderstood what you did